### ISSUE TITLE
[scopes] Defining scopes for all routes

### ISSUE THREAD

**Issue by frgfm** [pyronear/pyro-api#45]
Hey there :wave: 

Playing around with the PRs, I noticed that some routes are protected by scopes while others aren't. If some of those are correct, I just want to make sure we do consider carefully those choices for each router:

- [x] alerts: _should it be exposed to everyone on Read? admin & device for create, update, delete?_ (#99, #122)
- [x] devices: _"admin" by default for all routes, "me" added to self-ID routes_ (#121)
- [x] events: _should it be exposed to everyone on Read? admin & device for create, update, delete?_ (#59, #99)
- [x] installations: _should it be exposed to everyone on Read? admin only for create, update, delete?_ (#99, #120)
- [x] media: _we need some scope but can't be admin only, since a logged user whose scope includes a given device, should have access to its media_ (#99, #118)
- [x] sites: _should it be exposed to everyone on Read? admin only for create, update, delete?_ (#58, #99) 
- [x] users: _"admin" by default for all routes, "me" added to self-ID routes_ (#15, #44)

Those where I have no clue whether we should add some:
- [x] accesses: _"admin" by default for all routes_ (#104)
- [x] login: no scope

cc @florianriche @jeanpasquier75 @fe51 @martin1tab

**Comment by jeanpasquier75** [comment]
You are right we should protect the API 👍 
* login must not have a required scope -> everyone can request an access
* accesses aims to be removed (I guess) and is used so far for debugging but we should require an admin scope (to prevent exposing list of login/scopes)

**Comment by fe51** [comment]
I think we need to protect all routes expect login to request an access (thus for those you were suggesting, at least, to have any of the scopes ?)

- users: ok with the suggestion
- alerts: R for the owner of the device ? (and admin & device)
- devices: ok as it is (with some route with "device" scope)
- installations: CRUD for the owner of the device ? (and admin)
- media: CRUD for the owner of the device ? (and admin)

**Comment by frgfm** [comment]
So we would get the following :

router | Create | Read | Update | Delete
-- | -- | -- | -- | --
alerts | admin, device | None | admin, device | admin
devices | admin, user | admin, user | admin, user, self | admin, user
events | admin, device | None | admin, device | admin
installations | admin | None | admin | admin
media | admin, device | admin, device | admin, device | admin
sites | admin | None | admin | admin
users | admin | admin, self | admin, self | admin, self
accesses | NA | admin | NA | NA
login | None | NA | NA | NA


If that's alright with everyone, we have many PRs to open in order to handle this for the release :sweat_smile: 
Feel free to open one, only put a comment here saying which router you are taking care of just before starting the PR to avoid duplication!

**Comment by fe51** [comment]
I will start with routes sites

**Comment by fe51** [comment]
and events #59

**Comment by florianriche** [comment]
I think we need to change the way the auth are designed.
I thought the list of scopes was " or " list of permissions but it is actually a "and".
E.G if the scopes are ["user", "admin"] inly the users that are both admin and user can access the routes.

If we look at `deps.py`:

```python3
for scope in security_scopes.scopes:
        if scope not in token_data.scopes:
```
We see that if one scope of the user is not in the scopes specified in the security, then it fails.
I would suggest to
1. either change that piece of code so that one matching scope is enough (preferred solution)
2. Create new scopes like" user_x_admin"; "user_x_device" ...

**Comment by fe51** [comment]
Good catch @florianriche ! 

I clearly prefer the first suggestion and I think it fits our needs ! I can make a quick PR to correct it !

**Comment by frgfm** [comment]
Lowering the priority of the issue since this shouldn't be blocking for the incoming release!

**Comment by fe51** [comment]
Following the last discussion about scopes, and following OpenAPI, we should go back on #64, and keep permission conditions as an "and" operator.

In this case, currently, we should have three possible scopes : "me", "devices", "admin", and a user can then accumulate scopes ( as an example, the current superuser could have the three current scopes).

 So, we would get the following CRUD routes :

router | Create | Read | Update | Delete
-- | -- | -- | -- | --
alerts | device | None | device | admin
devices | Me | Me | Me (device for some) | user ?
events | device | None | device | admin
installations | admin | None | admin | admin
media | device | device | device | admin
sites | admin | None | admin | admin
users | admin | Me | Me | admin
accesses | NA | admin | NA | NA
login | None | NA | NA | NA

Also, to avoid the superuser to be able to delete itself, shall we add a specific condition in `delete_user ` preventing the deletion of a user which has superuser as a login ? 

What do you think of this suggestion @florianriche @jeanpasquier75 @martin1tab @frgfm  ?

**Comment by frgfm** [comment]
@fe51 My personal issue with that would be the wrong assumption that there is a clear ranking between scopes. If we were having consecutively included scopes, taking the lowest acceptable for the route would work. But we diverged from that when we introduced both a `user` and `device` scopes.

Independently of the other scope an access has, if the route requires admin route, and the access has it, then that should be enough in my opinion. Thinking about what each scope means independently rather than cumulatively. Keeping the OR, means that each route is accessible to multiple groups of accesses, and I feel like it's a more natural way.

**Comment by florianriche** [comment]
I agree with @frgfm .
As long as we don't have an overly complex system, reasoning in terms of "roles" ("OR" scopes) is easier than in cumulative permissions.

**Comment by fe51** [comment]
Thank you for your answers @frgfm @florianriche and I agree with your points (no ranking between scopes and the fact that the actual system is easier), it is also more natural to me and I was afraid I was missing something !


Hence, I think #58 and #59 are juste waiting a last approval from a review to be merged :D

**Comment by fe51** [comment]
#### About alerts routes :

router | Create | Read | Update | Delete
-- | -- | -- | -- | --
alerts | admin, device | None | admin, device | admin

With "device" scope, one should be able to update an alert. However, it should only possible to update its alerts.
Thus, following the same approach of alerte creation(`create_alert` with scope "admin" and `create_alert_from_device` with scope "device"), it would imply to having function `update_alert_from_device` (router.put("/from-device" etc)). What do you think ? 

#### About devices routes

router | Create | Read | Update | Delete
-- | -- | -- | -- | --
devices | admin, user | admin, user | admin, user, self | admin, user

A user with scope "me", should be able to Read all routes for devices or only the devices it owns ?

Some similar thought for Update and Delete. Either a specific routes are created, either, a kind of if condition is added in the current functions according to the scope (if scopes "me", a check is done if the permission is okay). What do you think ? 



@florianriche @frgfm @jeanpasquier75 @martin1tab :)
After discussion, I would be happy to perform those implementations !

**Comment by jeanpasquier75** [comment]
> Some similar thought for Update and Delete. Either a specific routes are created, either, a kind of if condition is added in the current functions according to the scope (if scopes "me", a check is done if the permission is okay). What do you think ?

I agree on that point, we can do something like:
```
device = await crud.get_entry(devices, device_id)  # check device exists
user_entry = await crud.get(user.id, users)
if "admin" not in user_entry["scopes"] and device["owner_id"] != user.id:
    raise HTTPException(status_code=400, detail="Permission denied")
```

Moreover still on the devices routes, I figured out that:
* `PUT ​/devices​/{device_id}​/location (Update the location of a specific device)`, allowed for scopes "admin" and "me", performs the check on device owner and update only position.
* `PUT /devices/{device_id}/ (Update information about a specific device)`, allowed for scope "admin", does not perform any check and updates all fields: login, specs, position, last_ping and owner_id.

1. The proposal of checking only not admin would "align" the 2 routes on owner checking. 
2. But I think the 2nd route should only update "static" metadata fields (login, specs and owner_id) because an admin may not be aware of all the "changing" fields (position, last_ping). Or the next point could fix the problem:

If an optional field is not set, its value gets to null. Is it wanted in all cases ? There might be cases where we want to update only the fields that are specified in the request, no ? We can add on option to the base crud put or update_entry, what do you think ?

**Comment by frgfm** [comment]
@fe51 what do you mean exactly by "updating" its alert?
Considering that an alert is only a signal for start or end of an event (we are responsible to interpret them to have a consistent event table), an alert should only be updated for acknowledgement. Super admins may keep rights to edit them, but this should only be done for debugging purposes in my point of view.

For the devices routes, it should only be able to read information about devices she/he owns. I like @jeanpasquier75 suggestion which sums it up! And I agree that we need to align the owner checking on both routes you mentioned! 

Same about the rest, but this feels like a completely different topic (for field update with None), perhaps we should discuss this in another issue if you don't mind opening it @jeanpasquier75 :pray:

**Comment by frgfm** [comment]
@fe51 could you provide more explanations about your alert updating reference? (cf. my message above)

For the rest, we need to move on with an implementation decision and fix all these scopes shortly :pray: 
Perhaps aligning owner checking first as mentioned by @jeanpasquier75 and then open all the PRs to adjust all remaining scopes?

**Comment by fe51** [comment]
Hey there, 

About alerts, indeed, I agree admins rights are just allowed here for debugging purpose. 
I also agree that alerts should only be updated for acknowledgement, from an owner ? (or the account of a user related to an owner )

Yes @jeanpasquier75  is something I had in mind, nice suggestion !

**Comment by frgfm** [comment]
Well, I don't think so: a device creates an alert, and a user will acknowledge it. So someone that isn't either an admin or owner need to be able to update this.

Or we can consider that there are multiple owners :thinking:

**Comment by fe51** [comment]
Okay then, so we might need a new type of scope? 

Might be a solution

**Comment by frgfm** [comment]
Perhaps local_admin or something similar that will be added as a scope on corresponding routes?
Multiple owners is just pushing the problem back, so I doubt this is a good option

**Comment by frgfm** [comment]
Closing this issue since all routes are now correctly protected with the latest commits!
I opened up a discussion that is scope-related to handle user groups in #125
